redis-snapshot-replicator/main.tf

```diff
@@ -0,0 +1,258 @@
+resource "aws_iam_role" "replication-role" {
+  name = "redis-replication-role"
+
+  assume_role_policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "s3.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+POLICY
+}
+
+resource "aws_iam_policy" "replication-policy" {
+  name = "redis-replication-policy"
+
+  policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": [
+        "s3:GetReplicationConfiguration",
+        "s3:ListBucket"
+      ],
+      "Effect": "Allow",
+      "Resource": [
+        "${aws_s3_bucket.bucket.arn}"
+      ]
+    },
+    {
+      "Action": [
+        "s3:GetObjectVersion",
+        "s3:GetObjectVersionAcl"
+      ],
+      "Effect": "Allow",
+      "Resource": [
+        "${aws_s3_bucket.bucket.arn}/*"
+      ]
+    },
+    {
+      "Action": [
+        "s3:ReplicateObject",
+        "s3:ReplicateDelete"
+      ],
+      "Effect": "Allow",
+      "Resource": "${aws_s3_bucket.destination.arn}/*"
+    }
+  ]
+}
+POLICY
+}
+
+resource "aws_iam_policy_attachment" "replication" {
+  name       = "redis-replication-attachment"
+  roles      = ["${aws_iam_role.replication-role.name}"]
+  policy_arn = "${aws_iam_policy.replication-policy.arn}"
+}
+
+resource "aws_s3_bucket" "destination" {
+  bucket   = "redis-snapshots-destination"
+  region   = "eu-central-1"
+  provider = "aws.eu-central-1"
```

Comment:
the provider name should be something more generic, like `aws.replica`

Also, the providers in a module need to be declared somewhere. Take a look at the official Terraform documentation: https://www.terraform.io/docs/modules/usage.html#providers-within-modules

As an example, you can check the [`terraform-vault` module](https://github.com/skyscrapers/terraform-vault/blob/master/vault/main.tf)